File: 235.lowest-common-ancestor-of-a-binary-search-tree.java
/*
 * @lc app=leetcode id=235 lang=java
 *
 * [235] Lowest Common Ancestor of a Binary Search Tree
 */

// @lc code=start

import java.util.Stack;

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        int min = Math.min(p.val, q.val), max = Math.max(p.val, q.val);
        Stack<TreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.empty()) {
            TreeNode node = stack.pop();
            if (node != null) {
                if (node.right != null) stack.push(node.right);
                if (node.left != null) stack.push(node.left);
                stack.push(node); stack.push(null);
            } else {
                node = stack.pop();
                if (node.val >= min && node.val <= max) return node;
            }
        }
        return null;
    }
}
// @lc code=end

File: 235.lowest-common-ancestor-of-a-binary-search-tree.cpp
/*
 * @lc app=leetcode id=235 lang=cpp
 *
 * [235] Lowest Common Ancestor of a Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* lowestCommonAncestor(TreeNode* root, TreeNode* p, TreeNode* q) {
        if (root == nullptr) return nullptr;
        auto m_min = min(p->val, q->val), m_max = max(p->val, q->val);
        if (root->val > m_min && root->val < m_max) return root;
        if (root->val < m_min) return lowestCommonAncestor(root->right, p, q);
        if (root->val > m_max) return lowestCommonAncestor(root->left, p, q);
        return {};
    }
};
// @lc code=end
